File: LeetCode/22GenerateParentheses/GenerateParentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        # Backtracking -> Time:O(n), Space:O(h)
        stack = []
        ans = []
        def backtrack(open_count, closed_count):
            if open_count == closed_count == n:
                ans.append("".join(stack))
                return
            if open_count < n:
                stack.append("(")
                backtrack(open_count+1, closed_count)
                stack.pop()
            if closed_count < open_count:
                stack.append(")")
                backtrack(open_count, closed_count+1)
                stack.pop()
        backtrack(0, 0)
        return ans

File: LeetCode/22GenerateParentheses/GenerateParentheses.cpp
class Solution {
public:
    void backtrack(vector<string> &ans, int n, int open_count, int close_count, string curr) {
        if (open_count == close_count && open_count == n) {
            ans.push_back(curr);
            return;
        }
        if (open_count < n) {
            backtrack(ans, n, open_count + 1, close_count, curr + "(");
        }
        if (close_count < open_count) {
            backtrack(ans, n, open_count, close_count + 1, curr + ")");
        }
    }
    vector<string> generateParenthesis(int n) {
        // Backtracking -> Time:O(n), Space:O(h)
        vector<string> ans;
        backtrack(ans, n, 0, 0, "");
        return ans;
    }
};
